### PR TITLE
Profile send via ctl channel

### PR DESCRIPTION
**Pull request by sam-github**
connected to strongloop-internal/scrum-nodeops#602

### CODE REVIEW

**Comment by rmg** [comment]
How big can these profiler files get? They'll have to fit entirely in memory, at least twice, maybe more, on both the sending and receiving side of the control channel.

This could be a problem with memory constrained environments that don't have swap, like a small VM (t2 or t1 on AWS, or 512M or 1G on DO) or a docker container used for resource control.

**Comment by sam-github** [comment]
They can get really, really big. Particularly the heap snapshot. I'm not sure what to do about that. It needs to be transferred, its got to go over the websocket control channel. I'm not sure what we can do. The `readFile()` will fail if insufficient mem, I'm not sure about the entire channel, whether it will gracefully discard messages too large to be serialized.

Other than simply not support heap snapshots, its not clear if there are other options. ws doesn't support streaming data, unfortunately. We could implement a chunk-transfer protocol for very large messages... where the profile is sent back (or requested) in a series of smaller chunks... this might be the only way, long-term, but I'd like to ignore this problem until we have it...  and even the chunked transfer won't help if the FS is too small to store the data.

I considered adding a `heapSnapshot()` command to strong-agent, TakeHeapSnapshot is already covered by the compat layer, but then I realized I'd have to implement the OutputStream interface to build up a String or Buffer, which is a more than a few minutes work, particularly since node/v8 doesn't seem to have an efficient way to build up very large buffers if you don't know the size before hand.

And doing so would still leave the entire heap profile in memory.

The worst of it is, you are most likely to heap dump when you have lots and lots of memory in use due to a memory leak.

Any suggestions, @bnoordhuis @piscisaureus ?

**Comment by sam-github** [comment]
I take it back about `fs.readFile()` failing with insufficient memory... looks like the OOM killer nails you first, at least `fs.readFile("/dev/zero",` died with SIGKILL at about 87% mem usage (and 140% cpu).

**Comment by rmg** [comment]
We could either implement our own multiplexing and create a framing protocol for streaming the file in chunks or use an existing like the [substream plugin for primus](https://github.com/primus/substream) or similar.

**Comment by rmg** [comment]
Or a more generic stream module like https://github.com/dominictarr/mux-demux that might work with any of the control channel transports.

**Comment by sam-github** [comment]
Unless I miss something about those tools, they build on top of an existing stream interface, but ws doesn't implement a stream, so the stream tools don't work.

We'd have to first implement a stream across ws, that what I mean by implementing a "chunked transfer".

Seems like primus might implement streams already... which is the first primus feature I've heard of that might be relevant to our use case. Maybe we could rip that bit out. Its not a huge amount of work to add it ourself on top of ws, since it implements ordered reliable msg transfer.

**Comment by sam-github** [comment]
@slnode test please

**Comment by piscisaureus** [comment]
Maybe it's possible to just do an http POST request?

**Comment by sam-github** [comment]
POST should be possible, yes.

**Comment by sam-github** [comment]
@slnode test please

**Comment by rmg** [comment]
doh! @piscisaureus that's far too obvious..

**Comment by bnoordhuis** [comment]
@piscisaureus already suggested a good solution, but just in case:

> I considered adding a heapSnapshot() command to strong-agent, TakeHeapSnapshot is already covered by the compat layer, but then I realized I'd have to implement the OutputStream interface to build up a String or Buffer

It's not safe to call into V8 from inside `OutputStream::WriteAsciiChunk()`.  You either have to buffer the whole snapshot, write it to disk, or stream it over the network in C++ land.

**Comment by sam-github** [comment]
Posting it from C++ might be the long-term solution, then. Short-term is too ignore the problem, I think.

File: lib/debug.js [hunk 1]
```diff
@@ -1,6 +1,6 @@
 var debug = require('debug');
 var name = require('../package.json').name;
-var MAX = 80;
+var MAX = 120;
```
Comment (kraman) [inline lib/debug.js]: thats where it was hiding. i was looking for this setting :)

Comment (sam-github) [inline lib/debug.js]: I'm going to make a strong-debug wrapper... this code is cut-n-paste too many places.

**Comment by kraman** [comment]
LGTM

**Pull request merged**